zfit_physics/models/pdf_hypatia2.py

```diff
@@ -0,0 +1,330 @@
+from __future__ import annotations
+
+import numpy as np
+import tensorflow as tf
+import tensorflow_probability as tfp
+import zfit
+import zfit.z.numpy as znp
+from zfit import z
+from zfit.util import ztyping
+
+sq2pi = np.sqrt(2.0 * np.arccos(-1.0))
+sq2pi_inv = 1.0 / sq2pi
+logsq2pi = np.log(sq2pi)
+log_de_2 = np.log(2.0)
+
+
+@z.function(wraps="tensor")
+def bessel_kv_func(nu, x):
+    """Modified Bessel function of the 2nd kind."""
+    return tfp.math.bessel_kve(nu, x) / znp.exp(znp.abs(x))
+
+
+@z.function(wraps="tensor")
+def gammafunc(nu):
+    return znp.exp(tf.math.lgamma(nu))
+
+
+@z.function(wraps="tensor")
+def low_x_BK(nu, x):
+    return gammafunc(nu) * znp.power(2.0, nu - 1.0) * znp.power(x, -nu)
+
+
+@z.function(wraps="tensor")
+def low_x_LnBK(nu, x):
+    return znp.log(gammafunc(nu)) + (nu - 1.0) * log_de_2 - nu * znp.log(x)
+
+
+@z.function(wraps="tensor")
+def BK(ni, x):
+    nu = znp.abs(ni)
+    first_cond = tf.logical_and(x < 1.0e-06, nu > 0.0)
+    second_cond = tf.logical_and(tf.logical_and(x < 1.0e-04, nu > 0.0), nu < 55)
+    third_cond = tf.logical_and(x < 0.1, nu >= 55)
+    cond = tf.logical_or(first_cond, tf.logical_or(second_cond, third_cond))
+    return znp.where(cond, low_x_BK(nu, x), bessel_kv_func(nu, x))
```

Comment:
Slightly concerned about this one: Will the wrong numbers cause issues anywhere here, i.e. Nans in the bessel_kv_func?

The point is that both branches are evaluated, and while it's fine for the normal where, unfortunately the gradient can be affected by this (yep, annoying), as it's calculated by multiplying the not-used branch by 0. But this results again in NaNs.

Can we check the gradients of the function too? @ikrommyd we did that already in the past, didn't we? I just could not find the pdf, was it zfit?

Reply:
We didn't check gradients in ci. But I did that manually for the Faddeeva function. For PDFs so far we just knew that we weren't doing any operations that may cause this.

Reply:
Ah okay! Probably worth to add a check then

Reply:
<img width="751" alt="image" src="https://github.com/user-attachments/assets/04d89113-a624-4b78-9794-f25865999b6c">
So indeed there is a problem

Reply:
Yeah, it looks like there is a problem.
![image](https://github.com/user-attachments/assets/5f6a76de-5da9-45ea-811e-9660b7fbb9d3)

Reply:
Great, thanks a lot for checking! Okay, yep, I guess we need again "safe values", do you remember how that works? We need to first create an array of arguments where the unused arguments, the ones that should not be taken and evaluate to NaNs are actually set to something harmless that does not produce anything wrong (i.e. a small, constant value). The value we use doesn't matter as we won't use it later on, since we use the where, but it's just not to screw up the gradient (i.e. the boolean to decide what goes where is gonna be used two times (or more)

Reply:
There is the `safe_where` function in `zfit.z` as well no?

Reply:
I added safe wheres but it doesn't fix the problem. Gradients are still nan. See https://github.com/zfit/zfit-physics/pull/84/commits/3b95ae2dd335b1a490ca8425aea7d886919f3b0f

Reply:
Maybe there is another problem somewhere else. I'll try to check again tomorrow.

Reply:
@qishi-hep There are some issues with the gradients. They start from tfp probably because the Bessel kve doesn't have a gradient with respect to one parameter. There may probably be places elsewhere that make the gradiends go nan. We'll debug and see what we can do about it.